### PR TITLE
chore(ci-visibility): unshallow repository when git >= 2.27

### PR DESCRIPTION
**Pull request by romainkomorndatadog**
An unshallow command compatible with git >= 2.27 is now used when a shallow repository is detected.

Additionally, packfile upload is improved by searching for commits found in the repo that are not known to the ITR backend, rather than just HEAD.

For tests, a shallow clone was created using `git clone --depth=1 https://github.com/pallets/flask.git`, with git log showing only one result:
```
% git log --pretty=oneline
cb825687a592709f902f3d320d93987a0546fd28 (grafted, HEAD -> main, origin/main, origin/HEAD) Bump actions/checkout from 3.5.2 to 3.5.3 (#5186)
```

After:
```
cb825687a592709f902f3d320d93987a0546fd28 (HEAD -> main, origin/main, origin/HEAD) Bump actions/checkout from 3.5.2 to 3.5.3 (#5186)
51bf0fdd9018d63b7bc09db52a66b726c2c4bad8 Bump slsa-framework/slsa-github-generator from 1.6.0 to 1.7.0 (#5185)
0da5788efbda92b6b6239352c2b9a835efe0888f Bump dessant/lock-threads from 4.0.0 to 4.0.1 (#5184)
# ... ~20 lines
32d2f47ed1bf3f27fd46e1a54dd136688de9ab81 [pre-commit.ci] pre-commit autoupdate
c9b6110dec52ff483b42b1428221f903bf143788 (grafted) Merge branch '2.3.x'
7dbb2f7e05d2c0963e90016b1f7a3a45e98a865a (grafted) retarget pre-commit.ci
```

This is gated behind the unreleased Intelligent Test Runner (ITR) functionality, so there is no risk of unexpectedly unshallowing repositories.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Review by majorgreys — COMMENTED** [review]
Can these interface changes be kept internal or are they meant to be public?

**Comment by romainkomorndatadog** [comment]
> Can these interface changes be kept internal or are they meant to be public?

@majorgreys , I suppose the use of the shallow/unshallow code is private to our CI visibility client. Do  you want me to move the added functionality from `ext/git.py` to `internal/ci_visibility/git_client.py` ? I'm not opposed to that (calling `_git_subprocess_cmd()` from another module should work). I just felt following the existing pattern was reasonable.

The existing interface (`get_rev_list_excluding_commits()`) itself doesn't actually change. I could reimplement the new underlying function elsewhere but that doesn't feel DRY.

File: ddtrace/internal/ci_visibility/recorder.py [hunk 1]
```diff
@@ -273,6 +273,8 @@ def _fetch_tests_to_skip(self):
             self._test_suites_to_skip = []
             return
 
+        self._test_suites_to_skip = []
```
Comment (romainkomorndatadog) [inline ddtrace/internal/ci_visibility/recorder.py:276]: Moving this up prevents us from potentially crashing when fetching skippable tests fails.

Running more tests than expected (the consequence of failing to fetch skippable tests) is preferable to the recorder hitting an exception and preventing tests from running.

File: tests/ci_visibility/test_ci_visibility.py [hunk 2]
```diff
@@ -716,3 +715,75 @@ def test_civisibility_check_enabled_features_itr_enabled_malformed_response(_do_
...
+            _is_shallow_repository=classmethod(lambda *args, **kwargs: True),
+            _get_latest_commits=classmethod(lambda *args, **kwargs: ["latest1", "latest2"]),
+            _search_commits=classmethod(lambda *args: ["latest1", "searched1", "searched2"]),
+            _get_filtered_revisions=mock.DEFAULT,
+            _build_packfiles=mock.DEFAULT,
+            _upload_packfiles=mock.DEFAULT,
+        ):
+            with mock.patch.object(CIVisibilityGitClient, "_unshallow_repository") as mock_unshallow_repository:
+                # import pdb;
+                # pdb.set_trace()
```
Comment (gnufede) [inline tests/ci_visibility/test_ci_visibility.py:752]: Left behind comments

**Review by gnufede — DISMISSED** [review]

**Review by gnufede — APPROVED** [review]

File: ddtrace/ext/git.py [hunk 3]
```diff
@@ -120,13 +149,22 @@ def extract_latest_commits(cwd=None):
 
 
 def get_rev_list_excluding_commits(commit_shas, cwd=None):
+    return get_rev_list(excluded_commit_shas=commit_shas, cwd=cwd)
+
+
+def get_rev_list(excluded_commit_shas=None, included_commit_shas=None, cwd=None):
```
[the remark this answers is not in the document]
Comment (romainkomorndatadog) [reply ddtrace/ext/git.py]: > If this function is intended for end-users of the library, we should announce it in a release note and mark this as a feature change. But it seems this change is not intended to change the library interface, so I would make the new function internal.

Oh good. That actually answers a question I meant to ask. I wasn't sure if the new underlying method I was adding (`get_rev_list`) also needed to be public, given that `get_rev_list_excluding_commits` was becoming such a thin wrapper around it.

I'm following your suggestion and making it private.

**Review by gnufede — APPROVED** [review]

**Review by majorgreys — APPROVED** [review]

**Pull request merged**